site/_posts/2018-02-12-discovering-a-ruby-memory-leak.html.md

```diff
@@ -0,0 +1,441 @@
+
+Also, as mentioned, this is a very rare situation where the programming
+language, ruby, was actually at fault here (we can safely admit that the
+developers of ruby are WAY smarter than us, so this is generally a safe
+assumption).  Now, there is an assumed asterisk that goes along with this
+statement for a few reasons:
+
+* Most developers tend to place `require` statements at the top of files, so
+  they aren't triggered constantly like it was in our case.  We did this for
+  #Reasons™, but in short, this was mostly used to solve other deficiencies of
+  our monolith application that could have been solved in other ways.
+* The Rails guides tend to show using `Strings` when setting `autoload_paths`,
+  and so it is less likely developers will use "deferred requires" like this
+  and also have `Pathname` entries in their `$LOAD_PATH`.
+
+There is even a bit of a note in the ruby source of [`load.c`][12] that even
+mentions that non-string values in the load path are expanded every time and
+not cached, so it seems that using `Pathname` values in the `$LOAD_PATH` is
+somewhat discouraged.
+
+Bottom line:  Still kinda our fault in the end...
+
+* * *
+
+Well, at the very least, hopefully you found this blog post entertaining, and
+possibly you can learn from our successes, and TONS of failures, if you find
+yourself trying to unravel your own memory leaks or other weird performance
+bugs!  The key was to continually look at the issue from different
+perspectives, with various tools, and of course, sleep on it and come back at
+it with renewed vigor and a clear mind.
+
+A big thanks to Joe Rafaniello, Keenan Brock, Dennis Metzger, and Jason Frey
+for their contributions in helping to find this leak.  I happen to be the one
+documenting this experience (_**epic adventure**_\*\*\*\*\*\*), but it was
+definitely a team effort over many months.
+
+
+
+* \* For those unfamiliar with ManageIQ/CFME, these are queue based workers,
+  the first of which "Collects" metrics from an infrastructure provider (VMs,
+  Containers, Instances, etc.) managed by ManageIQ, and sends it to the
+  "Processor" in a generic fashion that it can be imported into our DB schema.
+* \*\* Many of the attempted (failed) scripts can be found [here][9]
+  https://github.com/NickLaMuro/miq_tools/tree/master/miq_server_leak_discovery
+* \*\*\* The `GC.start` calls here are to avoid typical memory usage of ruby
+  objects being allocated normally from improperly skewing the results.
+* \*\*\*\* The reason for these "deferred requires" is for base memory savings.
+  Since `linux_admin` is only needed by a small portion of our code base, but
+  adds around 5 to 10 MB when loading it, we only load it when we need it.
+* \*\*\*\*\* Note, ruby 2.5.0 might not leak but Pathnames are much slower in the
+  `$LOAD_PATH` than strings, perhaps, because they're not cached as mentioned
+  [here][12].
+* \*\*\*\*\*\* Over selling this quite a bit here... | (• ◡•)|/\\(❍ᴥ❍ʋ)
```

Comment:
Remove a ton of this. It's either not actually important *or* just try and put them more inline. For example

* The 'many failed attempts' could be just an inline link where ever you have this footnote. `After [many failed attempts](link), we did this...`
* The deferred requires bit seems worthless to know
* the GC.start one seems actually valid to have here as a technical footnote explanation.

Reply:
Addressed the first bullet here:

https://gist.github.com/NickLaMuro/730b252a48019cc6bef85270ac16ff44/revisions#diff-4756561d24941a282634640442f5d8be

Commit message comments:

First off, there was an extra link in that footnote that needed to go regardless, but there wasn't much added to the body of the post by just combining the footnote here, so I did that.  Tried to avoid adding a comma.

That said, I think the original suggestion for removing this footnote was thinking it existed in a different spot, so the suggestions there were not specifically followed.

Also, since we took the first sentence from the next paragraph and merged it into this first one, the formatting was updated to account for that.

Reply:
Regarding the rest of what was said here:

> * The deferred requires bit seems worthless to know

I challenge that notion since it was part of the reason we got into this mess, and is part of why this post is here.  The rational for it though, while I think has some use, doesn't really belong in the body of the post.

Doesn't look like we really address this point anywhere else in the article, but maybe I am wrong.

> the GC.start one seems actually valid to have here as a technical footnote explanation.

Good, didn't change. :D

For the others, did you just not specifically call those out, or do you think what was mentioned here was all that needed to be changed.  Besides the last footnote (which I am partial to for the ascii-moji alone), not sure I think they should be removed.  Would need more convincing.